architecting-cloud-native-dotnet-apps-for-azure/infrastructure-as-code.md

```diff
@@ -0,0 +1,91 @@
+---
+title: Infrastructure As Code
+description: Architecting Cloud Native .NET Apps for Azure | Infrastructure As Code
+ms.date: 06/30/2019
+---
+
+# Infrastructure As Code
+
+Cloud Native Applications tend to make use of all sorts of fantastic Platform as a Service component. On a cloud platform like Azure these components might include things like storage, Service Bus and the SignalR service. As applications become more complicated the number of these services in use is likely to grow. Just as how continuous delivery broke the traditional model of deploying to an environment manually the rapid pace of change also broke the model of having a centralized IT group manage environments. 
+
+Building environments can, and should, also be automated. There is a wide range of very well thought out tools which can make the process easy. 
+
+## Azure Resource Manager Templates
+
+Also known as ARM Template, Azure Resource Manager Templates are a JSON-based language for defining various resources in Azure. The basic schema looks something like this:
+
+```
+{
+  "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
+  "contentVersion": "",
+  "apiProfile": "",
+  "parameters": {  },
+  "variables": {  },
+  "functions": [  ],
+  "resources": [  ],
+  "outputs": {  }
+}
+```
+
+Within this template one might a storage container inside the resources section like so
+
+```
+"resources": [
+    {
+      "type": "Microsoft.Storage/storageAccounts",
+      "name": "[variables('storageAccountName')]",
+      "location": "[parameters('location')]",
+      "apiVersion": "2018-07-01",
+      "sku": {
+        "name": "[parameters('storageAccountType')]"
+      },
+      "kind": "StorageV2",
+      "properties": {}
+    }
+  ],
+```
+
+The templates can be parameterized so that one template can be reused with different settings to define develop, QA and production environments. This help eliminate surprises when migrating to a higher environment that is set up differently from the lower environments. The resources defined in a template are typically all created within a single resource group on Azure (it is possible to define multiple resource groups in a single ARM template but unusual). This makes it very easy to delete an environment by simply deleting the resource group as a whole. Cost analysis can also be run at the resource group level allowing for quick accounting of which environments are costing what.
```

Comment:
this help eliminate -> helps eliminate

Reply:
which environments are costing what -> how much each environment is costing or each environment costs